### ISSUE TITLE
PHP / HTML / CSS : Improvement suggestions

### ISSUE THREAD

**Issue by Ariane-B** [codota/TabNine#216]
Using TabNine while doing HTML/CSS/SCSS doesn't feel very miraculous. I have a feeling the algorithm was designed for "real" programming, and those are an afterthought. I feel they could use a bit more work. I'll list a few things I've noticed.

1. In CSS, I often use "shortcuts" of properties and then pressing Tab, which native Sublime expands into my desired property just fine. For instance, martop + Tab will expand into margin-top. bgcl + Tab will expand into background-color. TabNine's behavior regarding these is rather inconsistent. It doesn't always suggest the right properties, and if it does, they're pretty low in the suggestion list.

2. If I type, say, margin and TabNine suggests "margin-top", I press Tab and it expands into the property just fine. But let's say now I want to add a value. Let's say 30px. Well, to type 30px, I'll have pressed the 3 key. TabNine just changed my property to whatever the third suggestion was. That's certainly not ideal.

3. In HTML (well, PHP files, but outside of <?php ?> tags), TabNine just gets downright confused by Emmet. I use Emmet to expand, say, div.my-class>div.another-class into the proper tags and classes, but as I type, the suggestions TabNine makes are never very useful.

4. At work, we use spaced parentheses. For instance, my_function( param_1, param_2 ). Note the absence of a space after the function name, but the presence of space within the parentheses. TabNine gets rather confused by that. It'll sometimes add only the trailing space, and sometimes add none at all.

**Comment by scnerd** [comment]
Just another friendly user here, no relation to the T9 dev's. I'd like to note that suggestions 1, 3, and 4 are all already handled by current tools (Sublime, Emmet, and providing an auto-formatting configuration to your IDE). Not arguing that T9 couldn't maybe tweak things to better understand some rigid, long-term structures like HTML/XML have, but in general these behaviors are already well-defined and nicely handled by other tools, so I don't think they're interested in trying to provide alternative solutions to that problem.

Where TabNine really shines is predicting non-deterministic text that no static analysis could have known about. For example, when I'm writing repetitive documentation, it'll often suggest entire phrases or sentences that I can use as-is or with very minor tweaks. Lots of other great, similar examples.

Granted, as you suggest, I'm coming from a "real" programming direction (mostly Python), which definitely lends itself more to this kind of prediction. In this context, T9 works best as a complement to PyCharm's already excellent auto-complete features based on static analysis; if there's any way you can combine T9's suggestions with those provided by these other tools you mention, you'll probably be much more satisfied using each tool where it's strongest.

**Comment by Ariane-B** [comment]
Yes, I think you have a good idea. Sublime Text's native tools (mostly) have effective solutions to handle these sorts of cases. But using TabNine implies replacing Sublime's native autocomplete with TabNine. This has already been reported elsewhere, but TabNine also mostly disables Sublime snippets as a side effect of this, built-in or custom. 

And I agree that when doing actual programming (in a JavaScript file, or a long PHP code file, for instance), TabNine's suggestions can be downright magical. So if TabNine could find a way to better coexist with Sublime's native tools, or integrate them somehow, it would be ideal.

**Comment by cslycord** [comment]
I came across this problem as well. The first time I try entering css/html code, my atom editor would expand typing "div" into making an actual div container. Then after that, I changed its class to container as I needed.

After doing that, any time I type "div" and press tab, tabnine expands it to the literal string of text 'div class="container"', which both doesn't make an html object like I needed and threw in a default that I didn't want.
